Python/specialize.c

```diff
@@ -778,6 +793,156 @@ _Py_Specialize_StoreAttr(PyObject *owner, _Py_CODEUNIT *instr, PyObject *name, S
 }
 
 
+// Please collect stats carefully before and after modifying. A subtle change
+// can cause a significant drop in cache hits. A possible test is
+// python.exe -m test_typing test_re test_dis test_zlib.
+int
+_Py_Specialize_LoadMethod(PyObject *owner, _Py_CODEUNIT *instr, PyObject *name, SpecializedCacheEntry *cache)
+{
+    _PyAdaptiveEntry *cache0 = &cache->adaptive;
+    _PyAttrCache *cache1 = &cache[-1].attr;
+    _PyObjectCache *cache2 = &cache[-2].obj;
+
+    PyTypeObject *owner_cls = Py_TYPE(owner);
+    if (PyModule_CheckExact(owner)) {
+        int err = specialize_module_load_attr(owner, instr, name, cache0, cache1,
+            LOAD_METHOD, LOAD_METHOD_MODULE);
+        if (err) {
+            goto fail;
+        }
+        goto success;
+    }
+    if (owner_cls->tp_dict == NULL) {
+        if (PyType_Ready(owner_cls) < 0) {
+            return -1;
+        }
+    }
+    // Technically this is fine for bound method calls, but slightly slower at runtime.
+    if (Py_TYPE(owner_cls)->tp_dictoffset < 0) {
+        SPECIALIZATION_FAIL(LOAD_METHOD, SPEC_FAIL_NEGATIVE_DICTOFFSET);
+        goto fail;
+    }
+    PyObject **owner_dictptr = _PyObject_GetDictPtr(owner);
+    int owner_has_dict = (owner_dictptr != NULL && *owner_dictptr != NULL);
+    PyDictObject *owner_dict = owner_has_dict ? (PyDictObject *)*owner_dictptr : NULL;
+    // Check for classmethods.
+    int owner_is_class = (owner_cls->tp_getattro != PyObject_GenericGetAttr) &&
+        PyObject_TypeCheck(owner, &PyBaseObject_Type) && PyType_Check(owner);
+
+    PyObject *descr = NULL;
+    DesciptorClassification kind = 0;
+    if (owner_is_class) {
+        // class method
+        kind = analyze_descriptor((PyTypeObject *)owner, name, &descr, 0);
+        // Store the version right away, in case it's modified halfway through.
+        cache1->tp_version = ((PyTypeObject *)owner)->tp_version_tag;
+    }
+    else {
+        // instance method
+        kind = analyze_descriptor(owner_cls, name, &descr, 0);
+        cache1->tp_version = owner_cls->tp_version_tag;
+    }
+
+    if (kind != METHOD || descr == NULL) {
```

Comment:
Could you use a switch of the `kind` here, please? It makes it clearer what is being done for each kind, and you won't need a `#if SPECIALIZATION_STATS`.

Reply:
Switched to switch-case. But I still need the ``#if SPECIALIZATION_STATS`` because additional checks are required to differentiate the fail case that `analyze_descriptor` cannot do.

Reply:
Also if we're really micro-optimizing, making the switch have less cases may make it a smidgen quicker :). I vaguely recall that a single if-elif is still faster than a jump table, so I wouldn't want the compiler to generate a jump table just for the failure cases. Though at this point I'd be bikeshedding.

Reply:
> `analyze_descriptor` cannot do

Do you mean "cannot" or "does not". If you mean "does not", then we should add an extra value to `DesciptorClassification`

Reply:
Ah alright, understood. Thanks for explaining.

I meant "does not". I guess I'll add it to ``DesciptorClassification`` then :).